js/base/Exchange.js

```diff
@@ -660,18 +660,21 @@ module.exports = class Exchange {
         return new Promise ((resolve, reject) => resolve (Object.values (this.markets)))
     }
 
-    filterBySinceLimit (array, since = undefined, limit = undefined, key = 'timestamp', tail = false) {
```

Comment:
@samgermain can't we avoid this `tail = undefined` ? this is technically wrong if we want to keep `tail` as a bool (a bool can't be undefined/null) and will immediately be a syntax error in Typescript or another typed language.

Reply:
> @samgermain can't we avoid this `tail = undefined` ? this is technically wrong if we want to keep `tail` as a bool (a bool can't be undefined/null) and will immediately be a syntax error in Typescript or another typed language.

It's the same as making a number or string undefined. This will just need to be an optional parameter, which is possible in typescript and no we can't really avoid making it undefined, because the default behaviour depends on `since`. What if `False`/`True` is actually passed in? Then we want to make sure we use `False`/`True` and can't change it depending on if `since` is defined or not

Reply:
@samgermain 
> It's the same as making a number or string undefined

Strings might be undefined/null in most languages, but yes numbers cannot 

> which is possible in typescript, and no we can't really avoid making it undefined

possible does not mean correct, in JS/TS we can do a lot of weird stuff without proper types but that is basically perpetrating tech debt 

btw why should we tie here two apparently orthogonal concepts? (tail and since)

Reply:
> btw why should we tie here two apparently orthogonal concepts? (tail and since)

When since is given, we want to get items that come after since (the number of items we want is equal to limit)
When since is not given, we want to get the most recent items

Reply:
> possible does not mean correct, in JS/TS we can do a lot of weird stuff without proper types but that is basically perpetrating tech debt

Optionals are pretty common 

[Java](https://www.baeldung.com/java-optional)
[Go](https://pkg.go.dev/github.com/markphelps/optional)

are there any languages that you're specifically worried about?

Reply:
> > btw why should we tie here two apparently orthogonal concepts? (tail and since)
> 
> When since is given, we want to get items that come after since (the number of items we want is equal to limit) When since is not given, we want to get the most recent items

@samgermain I know, but that's not the problem currently, we always get the elements after the since because that filtering is made on line 665. The problem is that after the since, we might want the items from one end or the other end. 

Example if you have this list, and a limit of 2:

[1,2,4] (this list already contains only elements after since)

Tail = true -> [2,4]
Tail = false -> [1,2]

My point is, this fix does not seem "generic", because the correct value of `tail` depends on the order of elements returned by the exchange (ascending or descending).

Moreover,  I still don't understand why this is directly related to `since` because on line 668 the elements are already taking into consideration the since value if any and the `tail` logic does not depend on that, (unless I'm missing something).

@samgermain Let me know what you think pls

Reply:
> > possible does not mean correct, in JS/TS we can do a lot of weird stuff without proper types but that is basically perpetrating tech debt
> 
> Optionals are pretty common
> 
> * [Java](https://www.baeldung.com/java-optional)
> * [Go](https://pkg.go.dev/github.com/markphelps/optional)
> 
> are there any languages that you're specifically worried about?

The problem is not the optional parameters but the default value of the primitive types. For instance, in Java/C# one bool or integer can't be `null` by default.
with that said let's not worry too much about it now, it's out of scope

Reply:
> 

We would need to make an optional type (do you know what optionals are by the way?) that would be capital `Boolean` or something, and the type would be `null | boolean`. We're already going to need to do this with Integers and with the values inside `has` because those are `boolean | undefined` right now. There's multiple ways to do this in [Java](https://stackoverflow.com/questions/314800/best-way-to-define-true-false-unset-state), and [C#](https://stackoverflow.com/questions/5362437/assign-null-value-to-boolean-variable)

> Moreover, I still don't understand why this is directly related to since because on line 668 the elements are already taking into consideration the since value if any and the tail logic does not depend on that, (unless I'm missing something).

Right now, 

 - if since is not given and the exchange returns 119 items (1/day from jan 1 -> apr 30), but our limit is 89, then 
     - we return the 89 items that occurred first (those from jan 1 -> march 31) 
     - instead of the hundred most recent (from jan 20 -> april 30), 
     - so the 19 most recent items (april 11 -> 30) aren't  returned
     - the user wants the most recent items from jan 20 -> apr 30

- if since is given, (as jan 1 lets say), then 
    - the user wants the 89 items that come after jan 1 (from jan 1 -> apr 10)
    - and it would be incorrect to not return the items from jan 1 -> jan 19)

Reply:
With 
If the exchange returns
```
  symbol | fundingRate |     timestamp |                 datetime
-----------------------------------------------------------------
BTC/USDT |  0.00001711 | 1670688000001 | 2022-12-10T16:00:00.001Z
BTC/USDT |  0.00000929 | 1670716800011 | 2022-12-11T00:00:00.011Z
BTC/USDT |  0.00000873 | 1670745600007 | 2022-12-11T08:00:00.007Z
BTC/USDT |  0.00000692 | 1670774400011 | 2022-12-11T16:00:00.011Z
BTC/USDT |  0.00002406 | 1670803200000 | 2022-12-12T00:00:00.000Z
BTC/USDT |  0.00002638 | 1670832000004 | 2022-12-12T08:00:00.004Z
BTC/USDT |  -0.0000167 | 1670860800000 | 2022-12-12T16:00:00.000Z
BTC/USDT |  0.00003003 | 1670889600000 | 2022-12-13T00:00:00.000Z
BTC/USDT |  0.00001971 | 1670918400018 | 2022-12-13T08:00:00.018Z
BTC/USDT |      0.0001 | 1670947200011 | 2022-12-13T16:00:00.011Z
```

and we have a limit of 5, and no since, then we want to return

```
  symbol | fundingRate |     timestamp |                 datetime
-----------------------------------------------------------------
BTC/USDT |  0.00002638 | 1670832000004 | 2022-12-12T08:00:00.004Z
BTC/USDT |  -0.0000167 | 1670860800000 | 2022-12-12T16:00:00.000Z
BTC/USDT |  0.00003003 | 1670889600000 | 2022-12-13T00:00:00.000Z
BTC/USDT |  0.00001971 | 1670918400018 | 2022-12-13T08:00:00.018Z
BTC/USDT |      0.0001 | 1670947200011 | 2022-12-13T16:00:00.011Z
```

If we have a since of `1670688000001`, and a limit of 5, then the response that we return should be

```
  symbol | fundingRate |     timestamp |                 datetime
-----------------------------------------------------------------
BTC/USDT |  0.00001711 | 1670688000001 | 2022-12-10T16:00:00.001Z
BTC/USDT |  0.00000929 | 1670716800011 | 2022-12-11T00:00:00.011Z
BTC/USDT |  0.00000873 | 1670745600007 | 2022-12-11T08:00:00.007Z
BTC/USDT |  0.00000692 | 1670774400011 | 2022-12-11T16:00:00.011Z
BTC/USDT |  0.00002406 | 1670803200000 | 2022-12-12T00:00:00.000Z
```

Reply:
@samgermain sorry, I expressed myself very poorly, and you're right.

I do have a second question tho, let's say we call this method with `tail` = `false`, meaning that the exchange returns the data in descending order (biggest timestamp/newest entry first), and `since` is also defined.

Example with dummy values

`since = 1700000000`

```
timestamps
1900000000
1800000000
1700000000
1600000000

after since filter:

1900000000
1800000000
1700000000

```

in that case, we would return the head of the filtered list

```
1900000000
1800000000
1700000000
```

⬆️ the thing is, we didn't change the sorting, so we still return using the descending order (newest first), which is not unified.

Do we need to invert the result here?

Reply:
@samgermain can you check this comment pls

Reply:
> @samgermain can you check this comment pls

Should we just get rid of the tail argument? I'm not sure why it's there, because it's not used, and then we could just sort all the records so that they're in the same order and return the first bunch if `since` is defined, and return the last bunch if `since` isn't defined

Reply:
> Should we just get rid of the tail argument? I'm not sure why it's there, because it's not used, and then we could just sort all the records so that they're in the same order and return the first bunch if since is defined, and return the last bunch if since isn't defined

@samgermain `tail` is needed and it's used to identify exchanges that return "inverted" data (newest entry first/ biggest timestamp first) hence my question, I think we might also need to invert the date in those cases to return them in a unified order (oldest/ smallest timestamps first)

Reply:
> > Should we just get rid of the tail argument? I'm not sure why it's there, because it's not used, and then we could just sort all the records so that they're in the same order and return the first bunch if since is defined, and return the last bunch if since isn't defined
> 
> @samgermain `tail` is needed and it's used to identify exchanges that return "inverted" data (newest entry first/ biggest timestamp first) hence my question, I think we might also need to invert the date in those cases to return them in a unified order (oldest/ smallest timestamps first)

Is this done for efficiency? We could just do a sort inside this method, or we could go through and find every exchange returning data in the order of "biggest timestamp first" and add the corresponding correct tail values